src/vos/vos_dtx.c

```diff
@@ -1096,8 +1098,30 @@ vos_dtx_check_shares(struct dtx_handle *dth, struct vos_dtx_entry_df *dtx,
 		return dtx_inprogress(dtx, 5);
 	}
 
-	if (dth == NULL)
-		return dtx_inprogress(dtx, 6);
+	/* Here, if the obj/key has 'prepared' DTX, but current @dth is NULL,
+	 * then it is the race case between normal IO and rebuild: the normal
+	 * IO creates the obj/key before the rebuild request being handled.
+	 *
+	 * Under such case, we should (partially) commit the normal DTX with
+	 * the shared target (obj/key) to guarantee the rebuild can go ahead.
+	 */
+	if (dth == NULL) {
```

Comment:
probably we need some kind of assertion to make sure dth == NULL is the rebuild case.